src/MelodyGenerator.js

```diff
@@ -3,18 +3,122 @@ const MelodyQueue = require('./MelodyQueue.js');
 const waveGenerator = require('./WaveGenerator.js');
 const note = require('./note.js');
 
-class MelodyGenerator
-{
+class MelodyGenerator {
     fromString(melodyString) {
         // put your code here
+        this.defines = [];
+        this.sounds = [];
+        let step1 = melodyString.replace(new RegExp(', ', "g"), ',').split("\n");
+
+        for (let i = 0; i < step1.length; i++) {
+            this.parse(step1[i]);
+        }
+
         let melody = new MelodyQueue();
-        melody.enqueueTone(durations.Q, note('A4'));
+        this.buildMelody(melody, this.sounds);
+
         return melody;
     }
 
     getNotesArray(notes) {
         return notes.map((n) => note(n));
     }
+
+
+    //pare a line
+    parse(line) {
+        let command = line.split(' ');
+
+        switch (command[0]) {
+            case 'play':
+                this.parseCommandPlay(command);
+                break;
+            case 'define':
+                this.parseCommandDefine(command);
+                break;
+
+            default:
+                break;
+        }
+
+    }
+
+    //parse notes and repeats
+    parseCommandPlay(comm) {
+        for (let i = 1; i < comm.length; i++) {
+            let cm = comm[i];
+            if (cm === 'repeat') {
+                let times = parseInt(comm[i + 1]);
+                for (let n = 0; n < times; n++) {
+                    this.add(comm[i + 3]);
+                }
+                i += 3;
+            } else {
+                if (comm[i] !== '|') this.add(comm[i]);
+            }
+        }
+    }
+
+    //parse definitions (aka macros)
+    parseCommandDefine(comm) {
+        let name = comm[1];
+
+        for (let i = 3; i < comm.length; i++) {
+            if (i % 2) {
```

Comment:
what.

Reply:
What 'what' ? :)

Reply:
What if code is wrong like `note(A,B) || note(A,B)`?

Reply:
Initial task was for pure code. Sure there is more to parse if not. :) Pipes more spaces in-between. I have solution for next task. Sure thing :)

Reply:
Hi Greg, one small tip: you can traverse through odd numbers only via:
`for (let i = 3; i < comm.length; i += 2)`

Reply:
Hi Tom, of course :)